enhancements/installer/azure-support-known-cloud-environments.md

```diff
@@ -0,0 +1,231 @@
+---
+title: azure-support-known-cloud-environments
+authors:
+  - "@abhinavdahiya"
+reviewers:
+  - TBD
+creation-date: 2020-05-11
+last-updated: 2020-05-11
+status: implementable
+see-also:
+  - "/enhancements/installer/aws-custom-region-and-endpoints.md"  
+replaces:
+superseded-by:
+---
+
+# Azure Support Known Cloud Environments
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As an administrator, I would like to deploy OpenShift 4 clusters to non-public Azure clouds. There are various known Azure clouds in addition to the `AzurePublicCloud` available to users, including but not limited to `AzureUSGovernmentCloud` etc. and requires the cluster to use different Azure API endpoints. The user should be allowed to provide the name of the Azure cloud and the installer and the cluster operators should use the correct API endpoints for that provided cloud.
+
+## Motivation
+
+### Goals
+
+1. Install cluster to known Azure clouds.
+
+### Non-Goals
+
+1. Supporting custom Azure clouds deployments that are not defined in Azure SDKs, i.e. custom BaseURI for endpoints will not suported.
+2. Supporting self-signed server certificates for endpoints specifically.
+
+## Proposal
+
+The installer will allow the users to specify the name of the Azure cloud using the install-config.yaml. The value will be restricted to one of the known [names](https://github.com/Azure/go-autorest/blob/9132adfc9db3007653dccb091b6a6c8e09b74ef5/autorest/azure/environments.go#L34-L39). When no cloud is specified the installer will default to `AzurePublicCloud`. The installer will use the endpoints pre-defined in the Azure SDK for the specified clouds to communicate with Azure APIs to perform validations, create machineset objects, and configuring the terraform `azurerm` provider.
+
+When using different Azure cloud, the cluster operators also need the discover the information, therefore, the installer will make these available using the config.openshift.io/v1 Infrastructure object. The cluster operators can then use Azure cloud name from the object to configure the Azure SDK to use the corresponding endpoints for communications.
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### US1: Installer cluster in MAG
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Infrastructure global configuration
```

Comment:
cc @openshift/api-reviewers 

since this recommends a change to config/v1 object.

Reply:
ping @deads2k @smarterclayton @derekwaynecarr